File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    map<char, string> createAlphabetDictionary() {
        map<char, string> alphabetDict;
        char currentChar = 'a';
        for (char i = '2'; i <= '9'; ++i) {
            int charCount = (i == '7' || i == '9') ? 4 : 3;
            string chars = "";
            for (int j = 0; j < charCount; ++j, ++currentChar) {
                chars += currentChar;
            }
            alphabetDict[i] = chars;
        }
        return alphabetDict;
    }

    void backtrackPerm(const string& digits, int depth, string& current, vector<string>& result, map<char, string>& alphaDict) {
        if (depth == digits.length()) {
            result.push_back(current);
            return;
        }
        char digit = digits[depth];
        for (char ch : alphaDict[digit]) {
            current += ch;
            backtrackPerm(digits, depth + 1, current, result, alphaDict);
            current.pop_back();
        }
    }

    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        map<char, string> alphaDict = createAlphabetDictionary();
        vector<string> result;
        string current;
        backtrackPerm(digits, 0, current, result, alphaDict);
        return result;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution(object):
    def create_alphabet_dictionary(self):
        alphabet_dict = {}
        current_char_code = ord('a')  
        for i in range(2, 10):
            char_count = 4 if i in [7, 9] else 3
            alphabet_dict[str(i)] = [chr(current_char_code + j) for j in range(char_count)]
            current_char_code += char_count
        return alphabet_dict

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits : return []
        alpha_dct = self.create_alphabet_dictionary()
        result, used = [], [[0]*len(alpha_dct[i]) for i in digits]
        def backtrack_perm(depth, arr, result) :
            if len(arr) == len(digits) :
                result.append(''.join(arr[:]))
                return
            for i in range(len(alpha_dct[digits[depth]])) :
                if not used[depth][i] :
                    used[depth][i] == True
                    arr.append(alpha_dct[digits[depth]][i])
                    backtrack_perm(depth+1, arr, result)
                    arr.pop()
                    used[depth][i] == False

        backtrack_perm(0, [], result)
        return result
